### ISSUE TITLE
[Question] Classification using Machine Learning Algorithms

### ISSUE THREAD

**Issue by whysetiawan** [leondgarse/Keras_insightface#124]
Hi, thanks for this awesome project. I have mobilefacenet and mobilenetv3 small models with GDC output. I'm struggling to use Classification Algorithms like SVM, KNN etc. I already tried other method like cosine similarity which is works really well. Do you have any advice to classify using SVM, KNN etc?

**Comment by leondgarse** [comment]
I only tried that a long time ago. Here's just my basic test, mostly it's fit -> precidt:
```py
import numpy as np
data = np.random.randn(1000, 512).astype(np.float32)  # fake known face embeddings as ground truth

from sklearn.svm import SVC
model = SVC()
model.fit(data, np.arange(data.shape[0]))  # fit to known face embeddings

model.predict(data[:10])  # Predict
# array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
```

**Comment by whysetiawan** [comment]
Thanks for the fast response, do i need to normalize the data first using `sklearn.preprocessing.normalize`?

**Comment by leondgarse** [comment]
Yes, it needs to be normalized, also before `predict`. Here's some results I previously tested `comparing nmslib / dot / svm dist calculation` if you are interested:
```py
!pip install nmslib
import nmslib
data = np.random.randn(1000, 512).astype(np.float32)
index = nmslib.init(method='hnsw', space='cosinesimil')
index.addDataPointBatch(data)
%time index.createIndex({'post': 2}, print_progress=True)
# CPU times: user 520 ms, sys: 27.8 ms, total: 548 ms
# Wall time: 70.3 ms

neighbours = index.knnQueryBatch(data[:10], k=1, num_threads=4)

from sklearn import metrics
from sklearn.svm import SVC
model = SVC()
%time model.fit(data, np.arange(data.shape[0]))
# CPU times: user 2.44 s, sys: 15.1 ms, total: 2.46 s
# Wall time: 2.46 s

''' nmslib '''
%timeit index.knnQueryBatch(data[:10], k=1, num_threads=4)
# 306 µs ± 2.71 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

''' dot '''
%timeit np.argmax(np.dot(data, data[:10].T), 0)
# 46.5 µs ± 179 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

''' svm '''
%timeit model.predict(data[:10])
# 36.1 ms ± 422 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
```py
import nmslib
data = np.random.randn(100000, 512).astype(np.float32)
index = nmslib.init(method='hnsw', space='cosinesimil')
index.addDataPointBatch(data)

%time index.createIndex({'post': 2}, print_progress=True)
# CPU times: user 20min 7s, sys: 667 ms, total: 20min 7s
# Wall time: 1min 42s

neighbours = index.knnQueryBatch(data[:10], k=1, num_threads=4)

''' nmslib '''
%timeit index.knnQueryBatch(data[:10], k=1, num_threads=4)
# 1.17 ms ± 2.06 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

''' dot '''
%timeit np.argmax(np.dot(data, data[:10].T), 0)
# 10.9 ms ± 14.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
